tests/suites/test_suite_psa_crypto.function

```diff
@@ -8091,3 +8651,315 @@ exit:
     PSA_DONE();
 }
 /* END_CASE */
+
+/* BEGIN_CASE depends_on:PSA_WANT_ALG_JPAKE */
+void ecjpake_setup( int alg_arg, int primitive_arg, int hash_arg, int role_arg,
+                    int input_first, data_t *pw_data,
+                    int expected_status_arg )
+{
+    psa_pake_cipher_suite_t cipher_suite = psa_pake_cipher_suite_init();
+    psa_pake_operation_t operation = psa_pake_operation_init();
+    psa_algorithm_t alg = alg_arg;
+    psa_algorithm_t hash_alg = hash_arg;
+    psa_pake_role_t role = role_arg;
+    mbedtls_svc_key_id_t key = MBEDTLS_SVC_KEY_ID_INIT;
+    psa_key_attributes_t attributes = PSA_KEY_ATTRIBUTES_INIT;
+    psa_status_t expected_status = expected_status_arg;
+    psa_status_t status = PSA_ERROR_CORRUPTION_DETECTED;
+    unsigned char *output_buffer = NULL;
+    size_t output_len = 0;
+
+    PSA_INIT( );
+
+    ASSERT_ALLOC( output_buffer,
+                  PSA_PAKE_OUTPUT_SIZE(alg, primitive_arg,
+                                       PSA_PAKE_STEP_KEY_SHARE) );
+
+    if( pw_data->len > 0 )
+    {
+        psa_set_key_usage_flags( &attributes, PSA_KEY_USAGE_DERIVE );
+        psa_set_key_algorithm( &attributes, alg );
+        psa_set_key_type( &attributes, PSA_KEY_TYPE_PASSWORD );
+        PSA_ASSERT( psa_import_key( &attributes, pw_data->x, pw_data->len,
+                                    &key ) );
+    }
+
+    psa_pake_cs_set_algorithm( &cipher_suite, alg );
+    psa_pake_cs_set_primitive( &cipher_suite, primitive_arg );
+    psa_pake_cs_set_hash( &cipher_suite, hash_alg );
+
+    PSA_ASSERT( psa_pake_abort( &operation ) );
+
+    TEST_EQUAL( psa_pake_set_user( &operation, NULL, 0 ),
+                PSA_ERROR_BAD_STATE );
+    TEST_EQUAL( psa_pake_set_peer( &operation, NULL, 0 ),
+                PSA_ERROR_BAD_STATE );
+    TEST_EQUAL( psa_pake_set_password_key( &operation, key ),
+                PSA_ERROR_BAD_STATE );
+    TEST_EQUAL( psa_pake_set_role( &operation, role ),
+                PSA_ERROR_BAD_STATE );
+    TEST_EQUAL( psa_pake_output( &operation, PSA_PAKE_STEP_KEY_SHARE,
+                                 NULL, 0, NULL ),
+                PSA_ERROR_BAD_STATE );
+    TEST_EQUAL( psa_pake_input( &operation, PSA_PAKE_STEP_KEY_SHARE, NULL, 0),
+                PSA_ERROR_BAD_STATE );
+
+    PSA_ASSERT( psa_pake_abort( &operation ) );
+
+    status = psa_pake_setup( &operation, &cipher_suite );
+    if( status != PSA_SUCCESS )
```

Comment:
Hmm let me check, I don't remember why I wrote this like that.

Reply:
Ok, it's designed to test failures on multiple calls, `psa_pake_setup()`, `psa_pake_set_role()` and first `psa_pake_input()/psa_pake_output()` depending on what we pass wrong in the test data.

I think we simply miss a final test at the end of `ecjpake_setup()` checking we should have failed before.

Reply:
Something like:
```diff
diff --git a/tests/suites/test_suite_psa_crypto.function b/tests/suites/test_suite_psa_crypto.function
index 4ca308d1a..b7f681e78 100644
--- a/tests/suites/test_suite_psa_crypto.function
+++ b/tests/suites/test_suite_psa_crypto.function
@@ -8786,6 +8786,8 @@ void ecjpake_setup( int alg_arg, int primitive_arg, int hash_arg, int role_arg,
                                         output_buffer, 66 ),
                         PSA_ERROR_BAD_STATE );
         }
+        else
+            goto exit;
     }
     else
     {
@@ -8819,8 +8821,13 @@ void ecjpake_setup( int alg_arg, int primitive_arg, int hash_arg, int role_arg,
                                          output_buffer, 512, &output_len ),
                         PSA_ERROR_BAD_STATE );
         }
+        else
+            goto exit;
     }
 
+    if( expected_status != PSA_SUCCESS )
+        TEST_ASSERT( ! "One of the previous calls should have returned the expected error." );
+
 exit:
     PSA_ASSERT( psa_destroy_key( key ) );
     PSA_ASSERT( psa_pake_abort( &operation ) );
```

Would make sure we don't miss an `expected_status`

Reply:
Ok, thanks for the explanation!

However I'm still not convinced this is what we want here. I feel like based on the arguments we know which step is supposed to fail, and if that's the case, then it would be better practice for the test for check that we indeed fail at the expected time. First thing that pops to mind is having separate `expected_status_setup`, `_role`, etc. arguments but that's a bit tedious (and perhaps not great for the readability of test data?). Second thought then: have an extra `failure_stage` argument, numerical, indicating what should fail (0 = nothing, 1 = setup, 2 = role, etc.)? Not ideal for readability either. (Or we can define constants in the `HEADER` portion of the `.function` file and use them in the `.data` file - I _think_.) Wdyt?

Of course if we really don't know in advance what function is supposed to return an error, then your approach is the right one and my point is moot - but then we should have a comment about it :)

Reply:
> Or we can define constants in the HEADER portion of the .function file and use them in the .data file

Yes, that (assuming you go with the which-function-fails strategy, which seems sensible to me).

Reply:
Thanks for the suggestions, I went on the path of the multiple status arguments since it's limited and we know exactly where it will fail.
I also figured out the `psa_pake_set_password_key()` wasn't properly tested, and it will be fixed.